zeshel/retrieval/model.py

```diff
@@ -0,0 +1,263 @@
+from transformers import BertModel, BertPreTrainedModel, BertConfig
+import torch
+import torch.nn as nn
+import copy
+from torch.nn import CrossEntropyLoss
+import torch.nn.functional as F
+import math
+
+NEAR_INF = 1e20
+
+
+class PolyAttention(nn.Module):
+    """
+    Implements simple/classical attention.
+    """
+
+    def __init__(self, dim=1, attn='cosine', residual=False, get_weights=True):
+        super().__init__()
+        if attn == 'cosine':
+            self.cosine = nn.CosineSimilarity(dim=dim)
+        self.attn = attn
+        self.dim = dim
+        self.get_weights = get_weights
+        self.residual = residual
+
+    def forward(self, xs, ys, mask_ys=None, values=None):
+        """
+        Compute attention.
+        Attend over ys with query xs to obtain weights, then apply weights to
+        values (ys if yalues is None)
+        Args:
+            xs: B x query_len x dim (queries)
+            ys: B x key_len x dim (keys)
+            mask_ys: B x key_len (mask)
+            values: B x value_len x dim (values); if None, default to ys
+        """
+        if self.attn == 'cosine':
+            l1 = self.cosine(xs, ys).unsqueeze(self.dim - 1)
```

Comment:
Oops, cosine similarity is not right here. The reason is the values are bounded in [-1, 1], so doing a softmax over them can never yield a peaky attention distribution. You have to do really weird tricks like adding extra parameters just to rescale (see Section 4.1 of Gillick et al.)! I'd suggest removing this option if you didn't find it helpful.